[build-webkit-org] Add steps for running smart pointer static analysis in Buildbot

briannafan opened the pull request:
#### dbbdbf11bdfce598329b5e0eca20741176550092
<pre>
[build-webkit-org] Add steps for running smart pointer static analysis in Buildbot
<a href="https://bugs.webkit.org/show_bug.cgi?id=272998">https://bugs.webkit.org/show_bug.cgi?id=272998</a>
<a href="https://rdar.apple.com/problem/126761949">rdar://problem/126761949</a>

Reviewed by Jonathan Bedard and Aakash Jain.

Add new steps that run static analysis with the --smart-pointer-only flag and organize the results for human consumption.

* Tools/CISupport/build-webkit-org/steps.py:
(ArchiveStaticAnalyzerResults): Changes output and dest paths.
(UploadStaticAnalyzerResults): Changes source path.
(ScanBuildSmartPointer): Runs build-and-analyze with --only-smart-pointers enabled.
(ParseStaticAnalyzerResults): Runs generate-dirty-files.py to get current lists of dirty files and UUIDs of bugs.
(CompareStaticAnalyzerResults): Runs compare-static-analysis-results.py to compare current and previous dirty files.
(UpdateSmartPointerBaseline): Updates baseline to current run&apos;s results.
(TransferToS3.doStepIf):
(ExtractTestResults):
(ExtractTestResults.resultDownloadURL): Add URL for downloading results.
(ExtractStaticAnalyzerTestResults): Modify paths for scan-build results.
(ExtractStaticAnalyzerTestResults.getLastBuildStepByName):
(ExtractStaticAnalyzerTestResults.addCustomURLs):

* Tools/CISupport/build-webkit-org/steps_unittest.py:
(TestScanBuildSmartPointer): Added.
(TestParseStaticAnalyzerResults): Added.
(TestCompareStaticAnalyzerResults): Added.
(TestUpdateSmartPointerBaseline): Added.

Canonical link: <a href="https://commits.webkit.org/278895@main">https://commits.webkit.org/278895@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d067b1ef069131a7a241fef1ba11990d8d523ec4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/51870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31182 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4234 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55138 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/2562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/54173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37557 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2272 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/53969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/28800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44743 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23327 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/2005 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2097 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56729 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/26991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/51888 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/28228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/44854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/29125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7581 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/27965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1565,6 +1565,201 @@ def getResultSummary(self):
+        self.commandFailed = False
+
+    def start(self):
+        self.command = '/bin/rm -rf {output_dir}; '.\
```
Why are we trying to do this as a giant shell command instead of a `steps.ShellSequence`?

briannafan replied:
Good point. I wrote this before I knew `steps.ShellSequence` existed :)

JonWBedard commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1565,6 +1565,201 @@ def getResultSummary(self):
+        return {u'step': status}
+
+
+class ParseStaticAnalyzerResults(shell.ShellCommand):
```
Can we use a `shell.ShellCommandNewStyle`, so we don't have to go back and update this class in the future when we update buildbot versions?

JonWBedard commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1565,6 +1565,201 @@ def getResultSummary(self):
+    resultMsg = ''
+
+    def start(self):
+        self.command = 'python3 Tools/Scripts/generate-dirty-files {scan_build_output} '.\
```
We usually only set up commands as single strings if we're piping the output to something locally. We're not doing that here, so this command should be a list.

JonWBedard commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1565,6 +1565,201 @@ def getResultSummary(self):
+        return {u'step': status}
+
+
+class CompareStaticAnalyzerResults(shell.ShellCommand):
```
Should use new-style shell command here too.

JonWBedard commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1565,6 +1565,201 @@ def getResultSummary(self):
+    descriptionDone = ['compared static analyzer results']
+    resultMsg = ''
+
+    command = 'python3 Tools/Scripts/compare-static-analysis-results '
```
It doesn't look like we're doing any piping, so this should be a list instead of a string.

JonWBedard commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1565,6 +1565,201 @@ def getResultSummary(self):
+        return {u'step': status}
+
+
+class UpdateSmartPointerBaseline(shell.ShellCommand):
```
This looks like it should be a `steps.ShellSequence`

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1565,6 +1565,201 @@ def getResultSummary(self):
+
+    def start(self):
+        self.command = '/bin/rm -rf {output_dir}; '.\
+            format(output_dir=os.path.join(self.getProperty('builddir'), 'build/scan-build-output'))
```
Let's try to use f-string instead of .format().
Earlier, I tried to convert all the existing .format() in this file in https://commits.webkit.org/272914@main

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1565,6 +1565,201 @@ def getResultSummary(self):
+
+    def __init__(self, **kwargs):
+        kwargs['timeout'] = 2 * 60 * 60
+        super(ScanBuildSmartPointer, self).__init__(**kwargs)
```
Nit: can skip class name in super() in python 3.

JonWBedard commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1568,6 +1585,215 @@ def getResultSummary(self):
+    resultMsg = ''
+
+    @defer.inlineCallbacks
+    def run(self):
```
Note that although you've built this like a new-style class, you're still inheriting from `shell.ShellCommand`

JonWBedard commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1594,7 +1820,7 @@ def run(self):
 
     def doStepIf(self, step):
-        return CURRENT_HOSTNAME in BUILD_WEBKIT_HOSTNAMES
+        return CURRENT_HOSTNAME in BUILD_WEBKIT_HOSTNAMES + TESTING_ENVIRONMENT_HOSTNAMES
```
I don't think we want to land this change generally.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1614,6 +1840,9 @@ def resultDirectoryURL(self):
         return self.resultDirectory.replace('public_html/', '/') + '/'
 
+    def resultDownloadURL(self):
+        return self.zipFile.replace('public_html/results/', '')
```
please verify if you want to remove /results as well from this url.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1625,6 +1854,22 @@ def run(self):
+
+    def addCustomURLs(self):
+        step = self.getLastBuildStepByName(CompareStaticAnalyzerResults.name)
+        step.addURL("View full static analyzer results", self.resultDirectoryURL() + "scan-build-output/results.html")
```
better to store this directory name `scan-build-output` in a variable instead of hard-coding in various places.

aj062 commented on Tools/CISupport/build-webkit-org/steps_unittest.py:
```diff
@@ -1836,3 +1838,163 @@ def test_skipped(self):
+
+        self.expectRemoteCommands(
+            ExpectShell(workdir=self.WORK_DIR,
+                        command=['/bin/sh', '-c', '/bin/rm -rf wkdir/build/scan-build-output'],
```
Nit: can use SCAN_BUILD_OUTPUT_DIR here as well.

briannafan committed:
[build-webkit-org] Add steps for running smart pointer static analysis in Buildbot
https://bugs.webkit.org/show_bug.cgi?id=272998
rdar://problem/126761949

Reviewed by Jonathan Bedard and Aakash Jain.

Add new steps that run static analysis with the --smart-pointer-only flag and organize the results for human consumption.

* Tools/CISupport/build-webkit-org/steps.py:
(ArchiveStaticAnalyzerResults): Changes output and dest paths.
(UploadStaticAnalyzerResults): Changes source path.
(ScanBuildSmartPointer): Runs build-and-analyze with --only-smart-pointers enabled.
(ParseStaticAnalyzerResults): Runs generate-dirty-files.py to get current lists of dirty files and UUIDs of bugs.
(CompareStaticAnalyzerResults): Runs compare-static-analysis-results.py to compare current and previous dirty files.
(UpdateSmartPointerBaseline): Updates baseline to current run's results.
(TransferToS3.doStepIf):
(ExtractTestResults):
(ExtractTestResults.resultDownloadURL): Add URL for downloading results.
(ExtractStaticAnalyzerTestResults): Modify paths for scan-build results.
(ExtractStaticAnalyzerTestResults.getLastBuildStepByName):
(ExtractStaticAnalyzerTestResults.addCustomURLs):

* Tools/CISupport/build-webkit-org/steps_unittest.py:
(TestScanBuildSmartPointer): Added.
(TestParseStaticAnalyzerResults): Added.
(TestCompareStaticAnalyzerResults): Added.
(TestUpdateSmartPointerBaseline): Added.

Canonical link: https://commits.webkit.org/278895@main
